kernel/syscalls.c

```diff
@@ -152,6 +153,73 @@ ulong sys_times(struct tms *user_buf)
    return (ulong) get_ticks();
 }
 
+static struct k_timeval 
+timespec_to_timeval(struct k_timespec64 ts)
+{
+   return (struct k_timeval) {
+      .tv_sec = ts.tv_sec,
+      .tv_usec = ts.tv_nsec / 1000,
+   };
+}
+
+int sys_getrusage(int who, struct k_rusage *user_buf)
+{
+   struct task *curr = get_curr_task();
+   struct k_rusage buf;
+   clock_t utime_ticks;
+   clock_t stime_ticks;
+   struct k_timespec64 utime;
+   struct k_timespec64 stime;
+
+   if (who == RUSAGE_CHILDREN)
+      // TODO: Resource usage of a process's children
+      //       isn't supported yet!
+      return -EINVAL;
+
+   if (who != RUSAGE_SELF && who != RUSAGE_THREAD)
+      return -EINVAL;
+
+   // Since there can only be one thread per process,
+   // RUSAGE_SELF and RUSAGE_THREAD have the same meaning.
+
+   disable_preemption();
```

Comment:
Not sure if disabling the preemption is enough. I guess that the ticks are touched in the IRQ 0 handler (timer). So, you need to disable the interrupts here.

Reply:
While writing this function I tried to mirror the implementation of `sys_times`, which only guards the access of the tick counts with `disable_preemption`/`enable_preemption`. Should `sys_times` change too or is there a reason I'm missing why they're not the same?

Reply:
@cozis Great observation! I quickly too a look at `sys_times()` and `sched_account_ticks()` which is called from the IRQ handler. sched_account_ticks() might not need to disable the interrupts because the IRQ timer won't nest in itself (there's a check), but for `sys_times()` an all the other places where the ticks are used, just preemption is disabled and that doesn't seem enough. Even if the a single 32-bit address can be accessed kind-of atomically on x86, accessing both `curr->ticks.total` and `curr->ticks.total_kernel` certainly it is not atomic. The timer IRQ handler can run between the two writes. If we want consistent results, better just disable the interrupts, here and in all the other places where those variables are accessed. sched_account_ticks() might be simply moved in the disabled interrupts block in the caller and assert that interrupts are disabled internally.